File: leetcode/50-powx-n.rb
# @param {Float} x
# @param {Integer} n
# @return {Float}
def my_pow1(x, n)
  cache = {0 => 1, 1 => x}
  n < 0 ? 1 / _pow(x, -n, cache) : _pow(x, n, cache)
end


def _pow(x, n, cache)
  if cache.include? n
    cache[n]
  else
    mid = n / 2
    cache[n] = _pow(x, n - mid, cache) * _pow(x, mid, cache)
  end
end


def my_pow(x, n)
  if n == 0
    1
  elsif n == 1
    x
  elsif n < 0
    1 / my_pow(x, -n)
  else
    my_pow(x*x, n / 2) * (n & 1 == 0 ? 1 : x)
  end
end

File: leetcode/50-powx-n.cpp
/*
Implement pow(x, n).
*/
class Solution {
public:
    double myPow(double x, int n) {
        double ans = 0;
        switch (n) {
        case -1:
            ans = 1/x;
            break;
        case 0:
            ans = 1;
            break;
        case 1:
            ans = x;
            break;
        default:
            auto tmp = myPow(x, n>>1);
            ans = tmp * tmp;
            if ((n & 1) == 1) {
                ans = ans * x;
            }
            break;
        }
        return ans;
    }
};
